adagrams/game.py

```diff
@@ -1,11 +1,168 @@
+    "V": 4,
+    "W": 4,
+    "X": 8,
+    "Y": 4,
+    "Z": 10,
+}
+
+
 def draw_letters():
-    pass
+
+    # Draw 10 letters randomly, considering the quantity limits
+    letter_list = list(LETTER_POOL.keys())
+    tiles = {}  # Dictionary with letter and count to avoid nested loops
+    letter_bank = [] 
+
+    while len(letter_bank) < 10:
+        random_number = random.randint(0, len(letter_list) - 1)
+        random_letter = letter_list[random_number]
+
+        if random_letter in tiles:
+            if tiles[random_letter] >= LETTER_POOL[random_letter]:
+                continue
+            tiles[random_letter] += 1
+        else:
+            tiles[random_letter] = 1
+            
+        letter_bank.append(random_letter)
+
+    return letter_bank
+
+
+# Count occurrences of a letter in an iterable
+def count_items(item, iterable):
+    count = 0
+    for x in iterable:
+        if x == item:
+            count += 1
+    return count
+
 
 def uses_available_letters(word, letter_bank):
-    pass
+    capital_word = word.upper()
+
+    for letter in set(capital_word):
+        if letter not in letter_bank or count_items(letter, capital_word) > count_items(
+            letter, letter_bank
+        ):
+            return False
+    return True
+
 
 def score_word(word):
-    pass
+    # Calculate score
+    score = 0
+    for letter in word.upper():
+        score += LETTERS_VALUE[letter]
+
+    # Add bonus points for words of length 7-10
+    if len(word) >= 7 and len(word) <= 10:
+        score += 8
+    return score
+
+
+# Function to find the maximum value in a list of numbers
+def calculate_max(iterable):
+    max = 0
+    for value in iterable:
+        if value > max:
+            max = value
+    return max
+
+
+# Function to find the minimum length of words in a list
+def calculate_min_length(list_of_words):
+    if not list_of_words:
+        return None
+
+    min_length = len(list_of_words[0])
+    for word in list_of_words:
+        if len(word) < min_length:
+            min_length = len(word)
+    return min_length
+
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    # Calculate scores of each word
+    word_score = {}
+    for word in word_list:
+        score = score_word(word)
+        word_score[word] = score
+
+    # Calculate max score
+    max_score = calculate_max(word_score.values())
+
+    # Find all candidate words with max scores
+    candidates = []
+    for word, score in word_score.items():
+        if score == max_score:
+            candidates.append(word)
+
+    # Find minimum length in candidates
+    min_length = calculate_min_length(candidates)
+
+    # Determine a winner considering rules
+    current_winner = ""
+    for word in candidates:
+        if len(word) == 10:
+            current_winner = word
+            break
+        if len(word) == min_length and not current_winner:
+            current_winner = word
+
+    return current_winner, max_score
```

Comment:
Overall this function works and passes all the tests, so great job! Specifically, this is a really great way to separate out each section of data and then parse through it one at a time. There is nothing inherently wrong with that and it's a great way to learn and make sure you understand each step that you are doing!

Once challenge I have for you is to remember that when it comes to these sorts of algorithms, it can be super helpful to keep in mind what information you need to store as well as what information you don't! For example, we really only need to hold onto words that could be or are the highest scoring word. We don't really care about any of the other words! The dictionary you create maps every word to a score but we likely won't need most of those scores and the words that map to them.

Your logic is super sound for how to discover which word and score to update and I think you could leverage that logic to limit the number of extra dictionaries and lists that you create here! No need to make the actual change, but it is something to think about moving forward!